lib/TopStoriesFeed.jsm

```diff
@@ -35,33 +35,64 @@ this.TopStoriesFeed = class TopStoriesFeed {
     this._prefs = new Prefs();
   }
 
-  init() {
-    const initFeed = () => {
-      SectionsManager.enableSection(SECTION_ID);
-      try {
-        const {options} = SectionsManager.sections.get(SECTION_ID);
-        const apiKey = this.getApiKeyFromPref(options.api_key_pref);
-        this.stories_endpoint = this.produceFinalEndpointUrl(options.stories_endpoint, apiKey);
-        this.topics_endpoint = this.produceFinalEndpointUrl(options.topics_endpoint, apiKey);
-        this.read_more_endpoint = options.read_more_endpoint;
-        this.stories_referrer = options.stories_referrer;
-        this.personalized = options.personalized;
-        this.show_spocs = options.show_spocs;
-        this.maxHistoryQueryResults = options.maxHistoryQueryResults;
-        this.storiesLastUpdated = 0;
-        this.topicsLastUpdated = 0;
-        this.domainAffinitiesLastUpdated = 0;
-
-        this.loadCachedData();
-        this.fetchStories();
-        this.fetchTopics();
-
-        Services.obs.addObserver(this, "idle-daily");
-      } catch (e) {
-        Cu.reportError(`Problem initializing top stories feed: ${e.message}`);
+  async onInit() {
+    SectionsManager.enableSection(SECTION_ID);
+    try {
+      const {options} = SectionsManager.sections.get(SECTION_ID);
+      const apiKey = this.getApiKeyFromPref(options.api_key_pref);
+      // Set this to true if we're not loading from cache.
+      let shouldBroadcast = false;
+      this.stories_endpoint = this.produceFinalEndpointUrl(options.stories_endpoint, apiKey);
+      this.topics_endpoint = this.produceFinalEndpointUrl(options.topics_endpoint, apiKey);
+      this.read_more_endpoint = options.read_more_endpoint;
+      this.stories_referrer = options.stories_referrer;
+      this.personalized = options.personalized;
+      this.show_spocs = options.show_spocs;
+      this.maxHistoryQueryResults = options.maxHistoryQueryResults;
+      this.storiesLastUpdated = 0;
+      this.topicsLastUpdated = 0;
+      this.storiesLoaded = false;
+      this.domainAffinitiesLastUpdated = 0;
+
+      // Cache is used for new page loads, which shouldn't have changed data.
+      // If we have changed data, cache should be cleared,
+      // and last updated should be 0, and we can fetch.
+      // Think there is a bug here.
+      await this.loadCachedData();
+      if (this.storiesLastUpdated === 0) {
+        shouldBroadcast = true;
+        await this.fetchStories();
       }
-    };
-    SectionsManager.onceInitialized(initFeed);
+      if (this.topicsLastUpdated === 0) {
+        shouldBroadcast = true;
+        await this.fetchTopics();
+      }
+
+      let updateProps = {};
+      if (this.stories) {
+        updateProps = {rows: this.stories};
+      }
+      if (this.topics) {
+        updateProps = {...updateProps, topics: this.topics, read_more_endpoint: this.read_more_endpoint};
```

Comment:
In this case you could just do `updateProps.rows = this.stories` and `Object.assign(updateProps, {topics: this.topics, read_more_endpoint: this.read_more_endpoint}`